lightning/src/util/persist.rs

```diff
@@ -64,41 +130,930 @@ impl<'a, A: KVStorePersister, M: Deref, T: Deref, ES: Deref, NS: Deref, SP: Dere
+						"ChannelMonitor was stored under the wrong key",
+					));
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to deserialize ChannelMonitor {}, reason: {}",
+					monitor_name.as_str(),
+					e,
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"Failed to deserialize ChannelMonitor",
+				))
+			}
+		}
+	}
+
+	/// Deserialize a channel monitor update.
+	fn deserialize_monitor_update(
+		&self,
+		monitor_name: &MonitorName,
+		update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let ns = self.monitor_update_namespace(monitor_name);
+		let key = update_name.as_str();
+		let update_bytes = self.kv_store.read(&ns, key)?;
+		match ChannelMonitorUpdate::read(&mut io::Cursor::new(update_bytes)) {
+			Ok(update) => Ok(update),
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to deserialize ChannelMonitorUpdate {}, reason: {}",
+					update_name.as_str(),
+					e,
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"Failed to deserialize ChannelMonitorUpdate",
+				))
+			}
+		}
+	}
+
+	/// Delete updates with an update_id lower than the given channel monitor.
+	fn delete_stale_updates<ChannelSigner: WriteableEcdsaChannelSigner>(
+		&self,
+		channel_id: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+	) -> io::Result<()> {
+		let monitor_name: MonitorName = channel_id.into();
+		let update_names = self.list_update_names(&monitor_name)?;
+		for update_name in update_names {
+			let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
+			if update.update_id != CLOSED_CHANNEL_UPDATE_ID
+				&& update.update_id <= monitor.get_latest_update_id()
+			{
+				let ns = self.monitor_update_namespace(&monitor_name);
+				let key = update_name.as_str();
+				self.kv_store.remove(&ns, key)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<ChannelSigner: WriteableEcdsaChannelSigner, K: Deref, L: Deref> Persist<ChannelSigner>
+	for MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	/// Persists a new channel. This means writing the entire monitor to the
+	/// provided [`KVStore`].
+	fn persist_new_channel(
+		&self,
+		funding_txo: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+		_update_id: MonitorUpdateId,
+	) -> chain::ChannelMonitorUpdateStatus {
+		// TODO: We really need a way for the persister to inform the user that
+		// its time to crash/shut down once these start returning failure. A
+		// PermanentFailure implies we should probably just shut down the node
+		// since we're force-closing channels without even broadcasting!
+		let monitor_name = match MonitorName::try_from(funding_txo) {
+			Ok(n) => n,
+			Err(_) => return chain::ChannelMonitorUpdateStatus::PermanentFailure,
+		};
+		let key = monitor_name.as_str();
+		let mut monitor_bytes = MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.to_vec();
+		monitor_bytes.extend(&monitor.encode());
```

Comment:
done, I think. this should at least end up with one contiguous allocation for the final value, not sure if `monitor.encode()` will actually allocate another Vec but that seems unavoidable if so.